File: java/mix/C04PrimeType4.java
public class C04PrimeType4 {
    public static boolean isPrime(int n) {
        if (n < 2) return false;
        if (n % 2 == 0) return n == 2;
        for (int d = 3; ; d += 2) {
            int q = n / d;
            if (q < d) break;
            if (n % d == 0) return false;
        }
        return true;
    }
}

File: cpp/mix/c04_prime_type4.cpp
#include <iostream>
using namespace std;

bool isPrime(int n) {
    if (n < 2) {
        return false;
    }
    if (n == 2 || n == 3) {
        return true;
    }
    if (n % 2 == 0 || n % 3 == 0) {
        return false;
    }
    int i = 5;
    while (i * i <= n) {
        if (n % i == 0 || n % (i + 2) == 0) {
            return false;
        }
        i += 6;
    }
    return true;
}

int main() {
    cout << boolalpha << isPrime(29) << endl; // Example usage
    return 0;
}
